common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -176,486 +174,120 @@ public final boolean recycle(T o, Handle<T> handle) {
-            elements[size] = null;
-            if (ret.lastRecycledId != ret.recycleId) {
-                throw new IllegalStateException("recycled multiple times");
-            }
-            ret.recycleId = 0;
-            ret.lastRecycledId = 0;
-            this.size = size;
-            return ret;
-        }
-
-        boolean scavenge() {
-            // continue an existing scavenge, if any
-            if (scavengeSome()) {
-                return true;
+                DefaultHandle<T> node = head.get();
+                if (node == null) {
+                    gcDetected(i);
+                    genHead[i] = NULL_REF;
+                    continue;
+                }
+                genHead[i] = node.next;
+                node.next = null;
+                return node;
             }
-
-            // reset our scavenge cursor
-            prev = null;
-            cursor = head;
-            return false;
+            return null;
         }
 
-        boolean scavengeSome() {
-            WeakOrderQueue prev;
-            WeakOrderQueue cursor = this.cursor;
-            if (cursor == null) {
-                prev = null;
-                cursor = head;
-                if (cursor == null) {
-                    return false;
-                }
-            } else {
-                prev = this.prev;
-            }
+        void gcDetected(int genIndex) {
+            int i = genIndex;
+            if (i > 0) {
+                // Promote this generation
 
-            boolean success = false;
-            do {
-                if (cursor.transfer(this)) {
-                    success = true;
-                    break;
-                }
-                WeakOrderQueue next = cursor.next;
-                if (cursor.owner.get() == null) {
-                    // If the thread associated with the queue is gone, unlink it, after
-                    // performing a volatile read to confirm there is no data left to collect.
-                    // We never unlink the first queue, as we don't want to synchronize on updating the head.
-                    if (cursor.hasFinalData()) {
-                        for (;;) {
-                            if (cursor.transfer(this)) {
-                                success = true;
-                            } else {
-                                break;
-                            }
-                        }
-                    }
+                int maxId;
+                if (i < genMaxId.length) {
+                    maxId = genMaxId[i];
+                } else {
+                    // Eden generation
 
-                    if (prev != null) {
-                        prev.setNext(next);
+                    maxId = parent.currentGenId.get();
+                    if (maxId == genMaxId[i - 1]) {
+                        // This is the first thread to detect the new generation
+                        parent.currentGenId.incrementAndGet();
```

Comment:
The check and the increment are not atomic, so isn't this racy?

Reply:
It'd be quite fair to use a `compareAndSet()` here to reduce mental burden and worries.

But I don't think the race matters _much_ as few `DefaultHandle()`s would be allocated within the race window. It mainly "wastes" an id. I do agree that any `DefaultHandles()`s allocated during the race window will be handled less than _ideal_, as they might not be treated as eden in some threads.

When writing the code, I thought of `currentGenId` as just a clock. The clock needs to tick frequently enough to separate generations, but if it runs too frequently the code would react reasonably. It was so long ago that I can't be certain, but I think I might have purposefully avoided CAS here to help prevent myself from assuming a nice, orderly progression to avoid overly relying on "clean" interactions between threads and make sure I handled "when things go wrong" cases.

Reply:
Ok. If it's fine for them race here, then I think it'd be good to call that out in a comment.